pkg/revisions/tag_watcher.go

```diff
@@ -0,0 +1,150 @@
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/pkg/log"
+)
+
+// TagWatcher keeps track of the current tags and can notify watchers
+// when the tags change.
+type TagWatcher interface {
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+	AddHandler(handler TagHandler)
+	GetMyTags() sets.Set[string]
+	GetAllTags() sets.Set[string]
+}
+
+// TagHandler is a callback for when the tags revision change.
+type TagHandler func(sets.Set[string])
+
+type tagWatcher struct {
+	revision string
+	handlers []TagHandler
+
+	queue           controllers.Queue
+	webhookInformer kclient.Client[*admissionregistrationv1.MutatingWebhookConfiguration]
+	mu              sync.RWMutex
+	tagsToRevisions map[string]string
+	revisionsToTags map[string][]string
+}
+
+func NewTagWatcher(client kube.Client, revision string) TagWatcher {
+	p := &tagWatcher{
+		revision:        revision,
+		mu:              sync.RWMutex{},
+		tagsToRevisions: map[string]string{},
+		revisionsToTags: map[string][]string{},
+	}
+	p.queue = controllers.NewQueue("tag", controllers.WithReconciler(p.updateTags))
+	p.webhookInformer = kclient.NewFiltered[*admissionregistrationv1.MutatingWebhookConfiguration](client, kubetypes.Filter{
+		ObjectFilter:    isTagWebhook,
+		ObjectTransform: kube.StripUnusedFields,
+	})
+	p.webhookInformer.AddEventHandler(controllers.ObjectHandler(p.queue.AddObject))
+
+	return p
+}
+
+func (p *tagWatcher) Run(stopCh <-chan struct{}) {
+	if !kube.WaitForCacheSync(stopCh, p.webhookInformer.HasSynced) {
+		log.Errorf("failed to sync tag watcher")
+		return
+	}
+
+	p.queue.Run(stopCh)
+}
+
+// AddHandler registers a new handler for updates to tag changes.
+func (p *tagWatcher) AddHandler(handler TagHandler) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.handlers = append(p.handlers, handler)
+}
+
+func (p *tagWatcher) HasSynced() bool {
+	return p.queue.HasSynced()
+}
+
+func (p *tagWatcher) GetMyTags() sets.Set[string] {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	result := sets.New(p.revisionsToTags[p.revision]...)
+	result.Insert(p.revision)
+	return result
+}
+
+func (p *tagWatcher) GetAllTags() sets.Set[string] {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	result := sets.New[string]()
+	for k, v := range p.tagsToRevisions {
+		result.InsertAll(k, v)
+	}
+	return result
+}
+
+// notifyHandlers notifies all registered handlers on tag change.
+// assumes externally locked.
+func (p *tagWatcher) notifyHandlers() {
+	myTags := p.GetMyTags()
+	for _, handler := range p.handlers {
+		handler(myTags)
+	}
+}
+
+func (p *tagWatcher) updateTags(key types.NamespacedName) error {
+	var revision, tagName string
+	wh := p.webhookInformer.Get(key.Name, "")
+	p.webhookInformer.Get(key.Name, "")
+	if wh != nil {
+		revision = wh.GetLabels()[label.IoIstioRev.Name]
+		tagName = wh.GetLabels()[tag.IstioTagLabel]
+	}
+	p.mu.Lock()
+	defer func() {
+		p.mu.Unlock()
+		p.notifyHandlers()
```

Comment:
this is the part I meant about notifyHandlers not being under a lock

Reply:
notifyHandlers calls GetMyTags, which acquires a readLock, so I will need to move locking logic into notifyHandlers.